### DISCUSSION TITLE
org-babel for xonsh

### DISCUSSION THREAD

**Discussion by jgarte** [xonsh/xonsh#5673] [Ideas]
Might be cool to have xonsh support for org-babel:

https://github.com/bzg/worg/blob/bb243c3cbc9c41e4fb445981a590688233127ac7/org-contrib/babel/languages/ob-doc-eshell.org#L19

Just a thought 💡

**Comment by anki-code** [comment]
Please provide more info about what is "org-babel" and example of the awesome use cases to motivate people to implement this. Creating new [xontrib](https://xon.sh/tutorial_xontrib.html) using [xontrib template](https://github.com/xonsh/xontrib-template) is forever welcome!

**Reply by jgarte** [reply]
Here's a motivating example:

https://orgmode.org/worg/org-contrib/babel/languages/ob-doc-eshell.html

**Comment by bestlem** [comment]
org-babel is a way of doing literate programming in emacs.

To add xonsh as a language for org-babel is by writing emacs code - hopefully there should be no need for any xonsh scripting just emacs-lisp.

**Reply by jgarte** [reply]
I know. But more importantly, what is needed is time to research further and implement it 😺

**Comment by WillForan** [comment]
There are some pointers on https://orgmode.org/worg/org-contrib/babel/languages/index.html#develop

For my immediate need, I was able to use a very bare implementation

```
;; create minimal package 'ob-xonsh'
(defun org-babel-execute:xonsh (body params)
  "Simplest execute wrapper. Does not handel sessions or parameters"
  (org-babel-eval "xonsh" body))

(provide 'ob-xonsh)

;; add to allowed languages
(org-babel-do-load-languages 'org-babel-load-languages '((xonsh . t)))
```

An easy route to a more fully functioning babel execute could be to layer on top of the xonsh jupyter kernel. Something similar for `dyalog` is explored on https://j-minster.github.io/org-babel-APL-update.html


Also hijacking ob-python could be fruitful. That was my first attempt but it wasn't as plug-and-play as I hoped. Below doesn't work.
```
(defun org-babel-execute:xonsh (body params)
  "Try to reuse python. Didn't work b/c return is appended to body?"
  (let ((org-babel-python-command "xonsh")) ;; swap python command for xonsh
     (org-babel-execute:python body params)))
```

**Comment by zeeros** [comment]
Support for `:session` and `:async` would be great, but so far I've managed to get `:var` working.

```
(defun org-babel-variable-assignments:xonsh (params)
  "Return a list of xonsh statements assigning variables from PARAMS."
  (mapcar (lambda (pair)
            (format "%s = %S" (car pair) (cdr pair)))
          (org-babel--get-vars params)))
(defun org-babel-execute:xonsh (body params)
  "Execute a block of Xonsh code with org-babel, supports :var but no :session."
  (let* ((vars (org-babel-variable-assignments:xonsh params))
         (full-body (string-join (append vars (list body)) "\n")))
    (org-babel-eval "xonsh" full-body)))
```
